lib/rex/post/meterpreter/channels/datagram.rb

```diff
@@ -0,0 +1,75 @@
+# -*- coding: binary -*-
+
+require 'rex/io/datagram_abstraction'
+require 'rex/post/meterpreter/channels/socket_abstraction'
+
+module Rex
+module Post
+module Meterpreter
+
+###
+#
+# Stream
+# ------
+#
+# This class represents a channel that is streaming.  This means
+# that sequential data is flowing in either one or both directions.
+#
+###
+class Datagram < Rex::Post::Meterpreter::Channel
+
+  include Rex::Post::Meterpreter::SocketAbstraction
+  include Rex::IO::DatagramAbstraction
+
+  class << self
+    def cls
+      return CHANNEL_CLASS_DATAGRAM
+    end
+  end
+
+  module SocketInterface
+    include Rex::Post::Meterpreter::SocketAbstraction::SocketInterface
+    def type?
+      'udp'
+    end
+
+    def recvfrom_nonblock(length,flags = nil)
+      return [super(length, flags)[0], super(length, flags)[0]]
```

Comment:
The remote end writes the gram, then does another write for the sockaddr so as not to impede on the buffer space for the gram. The native sockaddr structure comes from the C underbelly and we can't alter that behavior to return remote address info.